docs/proposals/horizontal-pod-autoscaler.md

```diff
@@ -0,0 +1,275 @@
+type ScaleStatus struct {
+	// Replicas is the number of actual replicas.
+	Replicas int
+
+	// Selector is a label query over pods that should match the replicas count.
+	Selector map[string]string
+}
+
+```
+
+Writing ```ScaleSpec.Count``` will resize the replication controller/deployment associated with
+the given Scale subresource.
+```ScaleStatus.Count``` will report how many pods are currently running in the replication controller/deployment,
+and ```ScaleStatus.PodSelector``` will return selector for the pods.
+
+## HorizontalPodAutoscaler Object
+
+We will introduce HorizontalPodAutoscaler object, it will be accessible under:
+
+```
+api/vX/horizontalpodautoscalers/myautoscaler
+```
+
+It will have the following structure:
+
+```go
+// HorizontalPodAutoscaler represents the configuration of a horizontal pod autoscaler.
+type HorizontalPodAutoscaler struct {
+	api.TypeMeta
+	api.ObjectMeta
+
+	// Spec defines the behaviour of autoscaler.
+	Spec HorizontalPodAutoscalerSpec
+
+	// Status represents the current information about the autoscaler.
+	Status HorizontalPodAutoscalerStatus
+}
+
+// HorizontalPodAutoscalerSpec is the specification of a horizontal pod autoscaler.
+type HorizontalPodAutoscalerSpec struct {
+	// ScaleRef is a reference to Scale subresource. HorizontalPodAutoscaler will learn the current
+	// resource consumption from its status, and will set the desired number of pods by modyfying its spec.
+	ScaleRef *SubresourceReference
+	// MinCount is the lower limit for the number of pods that can be set by the autoscaler.
+	MinCount int
+	// MaxCount is the upper limit for the number of pods that can be set by the autoscaler.
+	// It cannot be smaller than MinCount.
+	MaxCount int
+	// Target is the target average consumption of the given resource that the autoscaler will try
+	// to maintain by adjusting the desired number of pods.
+	// Currently this can be either "cpu" or "memory".
+	Target ResourceConsumption
+}
+
+// HorizontalPodAutoscalerStatus contains the current status of a horizontal pod autoscaler
+type HorizontalPodAutoscalerStatus struct {
+	// CurrentReplicas is the number of replicas of pods managed by this autoscaler.
+	CurrentReplicas int
+
+	// DesiredReplicas is the desired number of replicas of pods managed by this autoscaler.
+	// The number may be different because pod downscaling is someteimes delayed to keep the number
+	// of pods stable.
+	DesiredReplicas int
+
+	// CurrentConsumption is the current average consumption of the given resource that the autoscaler will
+	// try to maintain by adjusting the desired number of pods.
+	// Two types of resources are supported: "cpu" and "memory".
+	CurrentConsumption ResourceConsumption
+
+	// LastScaleTimestamp is the last time the HorizontalPodAutoscaler scaled the number of pods.
+	// This is used by the autoscaler to controll how often the number of pods is changed.
+	LastScaleTimestamp *util.Time
```

Comment:
:+1 this is also useful to prevent flapping (scale up/down in successive intervals).